Restore invalid cover image hashes

uklotzde opened the pull request:
Invalid cover image hashes with the value 0 are not handled correctly and wrong cover images are displayed. Instead those hashes need to be fixed and updated on-the-fly when the images are loaded the next time by `CoverArtCache`.

The updating is done in `CoverArtDelegate` that has access to `TrackModel`. `CoverArtCache` does not have direct access to `Track` objects or the library and I don't want to introduce such a dependency. Updating the hashes when tracks are loaded is too late, because first the images are loaded and displayed in the track table view without instantiating any `Track` objects. Not ideal but a clean solution that fits into our current design.

## Follow-up

PR #2524: SHA-256 digest and (background) color

uklotzde commented:
AutoDJProcessor tests are failing sometimes on CI. Inconvenient.

uklotzde commented:
Will be rebased on master after #2511 has been merged to fix the spurious test errors.

daschuer commented on src/library/coverart.cpp:
```diff
@@ -98,8 +97,8 @@ QImage CoverInfo::loadImage(
     } else if (type == CoverInfo::FILE) {
-        VERIFY_OR_DEBUG_ASSERT(!trackLocation.isEmpty()) {
-            kLogger.warning()
+        if (trackLocation.isEmpty()) {
```
please add a comment that explains why this can be empty.

uklotzde replied:
Done

daschuer reviewed (changes requested):
Thank you for the PR. I have left some comments, the manual test is pending.

uklotzde commented on src/library/coverart.cpp:
```diff
@@ -134,6 +133,34 @@ QImage CoverInfo::loadImage(
+                << *this;
+        // Preserve the track location, i.e. slicing during
+        // assignment is intended here!
+        static_cast<CoverInfoRelative&>(*this) = CoverInfoRelative();
```
Done

uklotzde commented:
Let's see if the tests will succeed after the database issues have been fixed.

uklotzde commented:
Still segfaults with ctest on Travis

uklotzde committed:
Extract guessing of cover art from CoverArtCache

uklotzde committed:
Get tracks by TrackRef from TableModel

uklotzde committed:
Fix spurious segfaults in library tests

uklotzde committed:
Prepare CoverArtCache for migration to new hash function

uklotzde committed:
Refresh cover art hashes when showing track table view

uklotzde committed:
Refresh cover art hashes when loading tracks

uklotzde commented:
After a long search for the spurious segfaults during tests I have rebuilt the whole branch and reorganized the commits into small, correlated work packages. Hopefully all tests will pass now.

uklotzde commented:
The remaining CI failures are now really unrelated.

uklotzde committed:
Merge branch 'master' into coverimagehash

uklotzde committed:
Merge branch 'master' into coverimagehash

uklotzde committed:
Fix more spurious test failures due caused by concurrent tasks

uklotzde committed:
Disable concurrent guessing of track cover art during tests

uklotzde commented:
@daschuer: Ready, finally.

The concurrent guessing of cover art was poorly documented and designed with lots of implicit assumptions and dependencies. It worked more or less by chance. This investigation and fix has cost me a LOT of time!

#2525 contains a separate fix for the application, independent of any test setup.

daschuer commented on src/library/coverartdelegate.cpp:
```diff
@@ -71,25 +77,41 @@ void CoverArtDelegate::slotOnlyCachedCoverArt(bool b) {
+            kLogger.info()
+                    << "Updating cover info of track"
+                    << coverInfo.trackLocation;
+            pTrack->setCoverInfo(coverInfo);
```
This is redundant if we pass the track into the request. 
Can't we do it at one place only?

uklotzde replied:
No, we don't. The CoverArtDelegate accesses query results directly. Track objects are only loaded if the CoverArtCache signals that the cover info has been updated.

daschuer reviewed (changes requested):
The API is still a bit cluttered. It would be nice to clean up. 
The manual test was successful. So it is already merge-able.

uklotzde commented on src/library/coverartcache.cpp:
```diff
@@ -26,47 +35,69 @@ inline QImage resizeImageWidth(const QImage& image, int width) {
-CoverArtCache::~CoverArtCache() {
-    qDebug() << "~CoverArtCache()";
+//static
+void CoverArtCache::requestCover(
```
I've just tried to preserve the original naming that used "request" for asynchronous invocations.

uklotzde replied:
request is always called with a pRequestor that expects to receive the result as a signal. What's wrong with this? Returning a QFuture from tryLoad would be nice, but it doesn't allow to return immediate results. Aynway, I'm not planning to rewrite this.

uklotzde commented on src/library/coverartutils.cpp:
```diff
@@ -181,10 +185,58 @@ CoverInfoRelative CoverInfoGuesser::guessCoverInfo(
+    }
+}
+
+void guessTrackCoverInfoConcurrently(
```
Concurrently, because it it is run by QtConcurrent. I've tried get this association into a readable function name.

daschuer replied:
Yes, I have been stumbled over the name and understood it after looking into the implementation. It first reads as there are concurrent guesses, which is not the case. 
It is just like "run a thread that guesses the cover".
This is no big issue for me but maybe we find a name that is less confusing.

uklotzde replied:
"run...blablabla....Guessing" is a bad name IMHO. It puts the the action that actually matters at the far end of the name.

daschuer replied:
Ok, so keep it as it is.

uklotzde committed:
Trim public API of CoverArtCache

Only support two distinct use cases, either with or without a Track
object at hand.

uklotzde committed:
Handle absolute/relative cover locations including edge cases

uklotzde commented on src/library/coverartcache.cpp:
```diff
@@ -26,47 +35,69 @@ inline QImage resizeImageWidth(const QImage& image, int width) {
-                 << requestInfo << pRequestor <<
-                desiredWidth << onlyCached << signalWhenDone;
+//static
+void CoverArtCache::requestTrackCover(
```
Done

uklotzde replied:
The _request_ functions are static and return nothing. Both (public) variants boil down to a single, special-case invocation of _tryLoad_. The _tryLoad_ function is non-static and may return something immediately, later, or never.

uklotzde replied:
The use cases are already split. How and why does this need to be split even further?

uklotzde committed:
Merge branch 'master' into coverimagehash

daschuer commented:
Ok, no need to overengineer this. It is already a good step forward.
Unfortunately build fails:

```
/home/appveyor/projects/mixxx/src/library/coverart.cpp:127:21: error: no match for ‘operator<<’ (operand types are ‘QDebug’ and ‘QFileInfo’)
             kLogger.warning()
             ~~~~~~~~~~~~~~~~~
                     << "loadImage"
                     ~~~~~~~~~~~~~~
                     << type
                     ~~~~~~~
                     << "cover does not exist:"
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~
                     << coverFile;
                     ^~~~~~~~~~~~
```

uklotzde commented:
> Ok, no need to overengineer this. It is already a good step forward.
> Unfortunately build fails:
> 
> ```
> /home/appveyor/projects/mixxx/src/library/coverart.cpp:127:21: error: no match for ‘operator<<’ (operand types are ‘QDebug’ and ‘QFileInfo’)
>              kLogger.warning()
>              ~~~~~~~~~~~~~~~~~
>                      << "loadImage"
>                      ~~~~~~~~~~~~~~
>                      << type
>                      ~~~~~~~
>                      << "cover does not exist:"
>                      ~~~~~~~~~~~~~~~~~~~~~~~~~~
>                      << coverFile;
>                      ^~~~~~~~~~~~
> ```

Backward compatibility with legacy Qt versions strikes once again.

uklotzde committed:
Fix compatbility issues with legacy Qt version

uklotzde committed:
Merge branch 'master' into coverimagehash

uklotzde commented:
Anything else to do? CI failures are unrelated.

uklotzde committed:
Merge branch 'master' into coverimagehash

uklotzde commented:
@daschuer Ping.

I have more changes ready on top of this PR that will allow to share UI code between the internal track collection and external track collections like aoide. Mainly additional base classes and a signal/slot optimization for the CoverArtDelegate. I already experienced inconsistencies due to redundant code that I want to prevent in the future.

uklotzde committed:
Merge branch 'master' into coverimagehash

daschuer commented:
No issues found during manual tests. LGTM. Thank you.
